This repository contains the code necessary to reproduce the analysis and figures for

*"Ancestry-based differences in the immune system are associated with lupus severity"* (2022), Slight-Webb, S, Thomas, K., et al...

# Installation

Download the code using

```git clone https://github.com/KevinTThomas/sle_activity_scrnaseq.git```

# Requirements

The analysis requires [R](https://www.r-project.org/) (>=4.1), 
[Rstudio](https://www.rstudio.com/) (>= 2021.09.0), 
[Python](https://www.python.org/) (>=3.8), 
and the packages as listed in the 
[R/packages.R](https://github.com/KevinTThomas/sle_activity_scrnaseq/blob/f41846b4833b6e66b6cc6c0d6ff57a1a2b15ebbe/R/packages.R)
file.

To improve reproducibility, ease installation, and prevent any "it works on my 
machine" effects, it is recommended that the analysis is run using the provided 
[Docker](https://docs.docker.com/get-docker/) container, which can be retrieved 
using

```docker pull milescsmith/sle_scrnaseq_da:4.1.1```

Dockerfiles for building the above image can be found in their own
[repository](https://gitlab.com/guthridge_informatics/control.git), specifically

Because of the size of the data, it is recommended that the analysis be run
on a Linux workstation with multiple cores and at least 64 GB of RAM or on an
institutional cluster.

# Analysis

The code is divided into two sections:

* An analysis pipeline that processes the counts generated by Cell Ranger, performs QC and background noise removal, generates a Seurat object, normalizes data, performs batch correction, and carries out clustering and dimensional reductions.

* Code for generating the figures used in the manuscript.

The code is implemented as a series of Rmarkdown files. Since the output of earlier notebooks is used as input for later notebooks, they should be run in the order:

1. 01_pp_soupx.Rmd
2. 02_pp_object_construction.Rmd
3. 03_pp_qc.Rmd
4. 04_analysis.Rmd
5. 05_plotting.Rmd

These can either be run from within Rstudio (using the `knit` function to compile the Rmarkdown documents into github-compatible markdown files or html files) or on the command line using:

```R -e "knitr::knit('analysis_pipeline/01_pp_soupX.Rmd')"```

to compile the first document, for example.

A small subset of the data is provided as a toy example in the `demo_data`
folder, which the code is currently written to analyze.

# Data

The scRNA-seq data has been deposited with the [Gene Expression Omnibus](https://www.ncbi.nlm.nih.gov/geo/)
and is available as series [GSE189050](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE189050)
